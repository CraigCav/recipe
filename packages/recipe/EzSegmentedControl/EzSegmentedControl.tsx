import React, {Fragment} from 'react';
import {Option, Fieldset} from './EzSegmentedControl.styles';
import Label from '../EzLabel';

const getId = (name, {value}) => `${name}-${value}`;

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

type Button = {
  value: string;
  label: string;
  disabled?: boolean;
};

type Props = Omit<React.FieldsetHTMLAttributes<any>, 'onChange'> & {
  active?: any;
  label: React.ReactNode;
  labelPosition: 'left' | 'hidden';
  name: string;
  options: Button[];
  onChange: (value: string) => void;
};

/**
 * EzSegmentedControls present mutually exclusive options as buttons.
 */
const EzSegmentedControl: React.FC<Props> = ({
  active,
  className,
  label,
  labelPosition,
  name,
  onChange,
  options,
}) => (
  <Fieldset className={className} role="radiogroup" aria-labelledby={`radiogroup-${name}`}>
    <Label id={`radiogroup-${name}`} position={labelPosition}>
      {label}
    </Label>
    {options.map(option => (
      <Fragment key={option.value}>
        <input
          id={getId(name, option)}
          type="radio"
          name={name}
          {...(onChange
            ? {checked: option.value === active}
            : {defaultChecked: option.value === active})}
          onChange={() => onChange && onChange(option.value)}
          disabled={option.disabled}
        />
        <Option key={option.value} htmlFor={getId(name, option)}>
          {option.label}
        </Option>
      </Fragment>
    ))}
  </Fieldset>
);

/**
 * defaultProps
 * @property {bool} active - No option is selected by default.
 * @property {string} labelPosition - The label is positioned on the left of the control by default.
 */
EzSegmentedControl.defaultProps = {
  labelPosition: 'left',
};

/**
 * @component
 */
export default EzSegmentedControl;
