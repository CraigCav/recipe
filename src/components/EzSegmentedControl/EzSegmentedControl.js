import React, {Fragment} from 'react';
import PropTypes from 'prop-types';
import {standard} from '../../themes';
import {Label, Legend, Fieldset} from './styles';

const getId = (name, {value}) => `${name}-${value}`;

/**
 * EzSegmentedControls present mutually exclusive options as buttons.
 */
const EzSegmentedControl = ({
  active,
  className,
  label,
  labelPosition,
  name,
  onChange,
  options,
  theme,
}) => (
  <Fieldset className={className} role="radiogroup" aria-labelledby={`radiogroup-${name}`}>
    <Legend id={`radiogroup-${name}`} labelPosition={labelPosition} theme={theme}>
      {label}
    </Legend>
    {options.map(option => (
      <Fragment key={option.value}>
        <input
          id={getId(name, option)}
          type="radio"
          name={name}
          defaultChecked={option.value === active}
          onChange={() => onChange && onChange(option.value)}
          disabled={option.disabled}
        />
        <Label theme={theme} key={option.value} htmlFor={getId(name, option)}>
          {option.label}
        </Label>
      </Fragment>
    ))}
  </Fieldset>
);

EzSegmentedControl.propTypes = {
  /**
   * The value that represents the active option.
   */
  active: PropTypes.oneOfType([PropTypes.any]),
  /**
   * Add label text describing the purpose of the group of options.
   */
  label: PropTypes.node.isRequired,
  /**
   * Optionally position (or visually hide) the label.
   */
  labelPosition: PropTypes.oneOf(['left', 'hidden']),
  /**
   * A unique name identifying the group of controls that make up the segmented control options.
   */
  name: PropTypes.string.isRequired,
  /**
   * An array of objects representing each button in the segmented control. Each object must have:
   * - value: the value to provide if this option is selected
   * - label: the label used to describe the option value
   */
  options: PropTypes.array.isRequired,
  /**
   * The theme controlling the default styles for the segmented control.
   */
  theme: PropTypes.shape({
    colors: PropTypes.object,
  }),
};

/**
 * defaultProps
 * @property {bool} active - No option is selected by default.
 * @property {stribg} labelPosition - The label is positioned on the left of the control by default.
 * @property {bool} theme - uses the standard theme by default.
 */

EzSegmentedControl.defaultProps = {
  active: undefined,
  labelPosition: 'left',
  theme: standard,
};

/**
 * @component
 */
export default EzSegmentedControl;
