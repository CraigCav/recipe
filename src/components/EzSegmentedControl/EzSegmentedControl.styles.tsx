import {css} from 'react-emotion';
import styled, {Themed} from '../../themes/styled';
import {shade} from 'polished';
import {hideVisually} from '../../styles';

const label = ({theme}: Themed) => css`
  cursor: pointer;
  background-color: white;
  border: 1px solid ${theme.colors.grays[400]};
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.12);
  color: ${theme.colors.grays[700]};
  display: block;
  flex: 1;
  padding: ${theme.spacing.xs} ${theme.spacing.md};
  text-align: center;
  margin-left: -1px;
  line-height: 1em;
  font-size: 0.85em;
  flex-basis: auto;

  &:first-of-type {
    margin-left: 0;
    border-top-left-radius: ${theme.borderRadius[1]};
    border-bottom-left-radius: ${theme.borderRadius[1]};
  }

  &:last-of-type {
    border-top-right-radius: ${theme.borderRadius[1]};
    border-bottom-right-radius: ${theme.borderRadius[1]};
  }

  &:hover {
    background-color: ${shade(0.97, 'white')};
    border-color: ${shade(0.85, theme.colors.grays[400])};
    color: ${shade(0.85, theme.colors.grays[700])};
  }

  input:checked + & {
    color: ${theme.colors.blues[500]};
    background: ${theme.colors.blues[200]};
    border: 1px solid ${theme.colors.blues[500]};
    position: relative;
  }

  input:focus + & {
    border-color: ${theme.colors.blues[600]};
    color: ${theme.colors.blues[600]};
  }

  input:disabled + & {
    cursor: default;
    opacity: 0.45;
    pointer-events: none;
  }
`;

export const Option = styled.label<Partial<Themed>>(label);

export const Fieldset = styled.div`
  display: flex;
  align-items: center;

  & input {
    ${hideVisually()};
  }
`;
