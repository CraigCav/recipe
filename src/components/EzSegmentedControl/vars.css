:root {
  --recipe-segmented-box-border-color: var(--recipe-alias-border-color);
  --recipe-segmented-box-background-color-hover: var(--recipe-global-color-static-gray-100);
  --recipe-segmented-box-border-color-hover: var(--recipe-global-color-static-gray-500);
  --recipe-segmented-box-border-radius: var(--recipe-alias-border-radius-regular);
  --recipe-segmented-box-text-color: var(--recipe-alias-text-color);
  --recipe-segmented-box-text-color-selected: var(--recipe-global-color-static-blue-700);
  --recipe-segmented-box-background-color-selected: var(--recipe-global-color-static-blue-200);
  --recipe-segmented-box-border-color-selected: var(--recipe-global-color-static-blue-500);
  --recipe-segmented-box-background-color-down: var(--recipe-global-color-static-gray-300);
}