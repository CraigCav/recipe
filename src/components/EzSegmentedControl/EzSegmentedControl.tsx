import React, {Fragment} from 'react';
import {Option, Fieldset} from './EzSegmentedControl.styles';
import Label from '../EzLabel';

const getId = (name, {value}) => `${name}-${value}`;

type Button = {
  value: string;
  label: string;
  disabled?: boolean;
};

type Props = React.FieldsetHTMLAttributes<any> & {
  active?: any;
  label: React.ReactNode;
  labelPosition: 'left' | 'hidden';
  name: string;
  options: Button[];
  onChange: (value: string) => void;
};

/**
 * EzSegmentedControls present mutually exclusive options as buttons.
 */
const EzSegmentedControl: React.SFC<Props> = ({
  active,
  className,
  label,
  labelPosition,
  name,
  onChange,
  options,
}) => (
  <Fieldset className={className} role="radiogroup" aria-labelledby={`radiogroup-${name}`}>
    <Label id={`radiogroup-${name}`} position={labelPosition}>
      {label}
    </Label>
    {options.map(option => (
      <Fragment key={option.value}>
        <input
          id={getId(name, option)}
          type="radio"
          name={name}
          defaultChecked={option.value === active}
          onChange={() => onChange && onChange(option.value)}
          disabled={option.disabled}
        />
        <Option key={option.value} htmlFor={getId(name, option)}>
          {option.label}
        </Option>
      </Fragment>
    ))}
  </Fieldset>
);

/**
 * defaultProps
 * @property {bool} active - No option is selected by default.
 * @property {string} labelPosition - The label is positioned on the left of the control by default.
 */
EzSegmentedControl.defaultProps = {
  labelPosition: 'left',
};

/**
 * @component
 */
export default EzSegmentedControl;
