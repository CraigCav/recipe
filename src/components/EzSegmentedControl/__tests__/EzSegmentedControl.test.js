import React from 'react';
import EzSegmentedControl from '../EzSegmentedControl';

describe('EzSegmentedControl', () => {
  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = render(
      <EzSegmentedControl
        name="test-segment"
        label="test segment"
        active="firstValue"
        options={[
          {label: 'first', value: 'firstValue'},
          {label: 'second', value: 'secondValue'},
          {label: 'third', value: 'thirdValue', disabled: true},
        ]}
      />
    );
    expect(actual).toMatchSnapshot();
  });

  /**
   * onChange tests.
   */
  describe('onChange', () => {
    it('is triggers onChange when segment is clicked', () => {
      const spy = jest.fn();

      const component = mount(
        <EzSegmentedControl
          name="test-segment"
          label="test segment"
          active="firstValue"
          onChange={spy}
          options={[
            {label: 'first', value: 'firstValue'},
            {label: 'second', value: 'secondValue'},
            {label: 'third', value: 'thirdValue'},
          ]}
        />
      );

      component
        .find('input')
        .last()
        .simulate('change', {target: {checked: true}});

      expect(spy).toHaveBeenCalledWith('thirdValue');
    });
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(
      <EzSegmentedControl
        name="test-segment"
        label="accessibility test"
        active="firstValue"
        options={[
          {label: 'first', value: 'firstValue'},
          {label: 'second', value: 'secondValue'},
          {label: 'third', value: 'thirdValue'},
        ]}
      />
    );
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
