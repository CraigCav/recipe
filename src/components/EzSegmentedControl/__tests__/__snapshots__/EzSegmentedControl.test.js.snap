// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EzSegmentedControl should render with default styles 1`] = `
.recipe-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.recipe-8 input {
  border: 0px;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.recipe-0 {
  color: #8B99A6;
  margin-right: 1rem;
}

.recipe-2 {
  cursor: pointer;
  background-color: white;
  border: 1px solid #CED4D9;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.12);
  color: rgba(0,0,0,0.6);
  display: block;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 0.75rem 1.5rem;
  text-align: center;
  margin-left: -1px;
  line-height: 1em;
  font-size: 0.85em;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
}

.recipe-2:first-of-type {
  margin-left: 0;
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
}

.recipe-2:last-of-type {
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
}

.recipe-2:hover {
  background-color: #f7f7f7;
  border-color: #afb4b8;
  color: rgba(0,0,0,0.94);
}

input:checked + .recipe-2 {
  color: #77b1e2;
  background: #EBF7FF;
  border: 1px solid #77b1e2;
  position: relative;
}

input:focus + .recipe-2 {
  border-color: #3E90D6;
  color: #3E90D6;
}

input:disabled + .recipe-2 {
  cursor: default;
  opacity: 0.45;
  pointer-events: none;
}

<div
  aria-labelledby="radiogroup-test-segment"
  class="recipe-8 recipe-9"
  role="radiogroup"
>
  <div
    class="recipe-0 recipe-1"
    id="radiogroup-test-segment"
  >
    test segment
  </div>
  <input
    checked=""
    id="test-segment-firstValue"
    name="test-segment"
    type="radio"
  />
  <label
    class="recipe-2 recipe-3"
    for="test-segment-firstValue"
  >
    first
  </label>
  <input
    id="test-segment-secondValue"
    name="test-segment"
    type="radio"
  />
  <label
    class="recipe-2 recipe-3"
    for="test-segment-secondValue"
  >
    second
  </label>
  <input
    disabled=""
    id="test-segment-thirdValue"
    name="test-segment"
    type="radio"
  />
  <label
    class="recipe-2 recipe-3"
    for="test-segment-thirdValue"
  >
    third
  </label>
</div>
`;
